logstash-core/src/main/java/org/logstash/Cloner.java

```diff
@@ -42,7 +42,8 @@ private Cloner(){}
         } else if (input instanceof List<?>) {
             return (T) deepList((List<?>) input);
         } else if (input instanceof RubyString) {
-            return (T) ((RubyString) input).doClone();
+            // new instance but sharing ByteList (until either String is modified)
+            return (T) ((RubyString) input).strDup(RubyUtil.RUBY);
```

Comment:
This is fantastic.

Looking at the jruby implementation, it appears that [`RubyString#dup()`](https://github.com/jruby/jruby/blob/9.2.11.1/core/src/main/java/org/jruby/RubyString.java#L811) may be marginally safer, since it bypasses the COW optimization when handling objects that merely _inherit_ `RubyString` (and therefore can't guarantee COW semantics), and relies instead on `RubyBasicObject#dup()` in those scenarios.

Reply:
thanks for the tip - that is ~ "breaking" change but definitely smt to consider.
went specifically for `strDup` as `doClone` also did not respect sub-classes.
would consider that a bug but the whole `Cloner` is not preserving types precisely.
(e.g. in theory there might be an ArrayList sub-class or a `TreeMap` sub-type)

Reply:
I wasn't terribly concerned about the cloned copy preserving the metaclass of the source object here, although both `RubyString#dup` and `RubyString#strDup` do so when only passed the runtime. 

My primary concern was objects that inherit `RubyString`, but do not fully implement Copy-On-Write semantics, resulting in a source object that could "infect" one of its clones. I doubt we have any such setup now, as it is generally considered an anti-pattern in ruby to subclass the core classes, and the setup would be even trickier because `RubyString.value` is private and after further inspection it appears that the implementation does a rock-solid job of not leaking references to the underlying object via any of its non-private methods.

You can dismiss my concern. This should be a great improvement to those who clone events in their pipelines.

Reply:
right `RubyString#strDup` preserves meta - unline `doClone`, was looking at the wrong piece.
anyway, glad we established that keeping the meta-class won't be of much concern.

> My primary concern was objects that inherit RubyString, but do not fully implement Copy-On-Write semantics

understood - seems to me it would be quite unusual. 
users (some gems) sub-class `String` in Ruby but it's still the same `RubyString` and things work, sub-classing in Java land is potentially an issue but so is doing: `str.to_java(IRubyObject).getByteList` (as you pointed out) and assuming write-ability.

this isn't a new concern as `RubyString`'s `ByteList` sharing happens under the hood already, from some of the string operations (`dup`, `to_sym`, `clear`) as an implementation detail.